### DISCUSSION TITLE
Ability to tell if query is from Core or ORM

### DISCUSSION THREAD

**Discussion by ggabriel96** [sqlalchemy/sqlalchemy#6830] [Usage Questions]
Hello there!

I have been studying and playing around with SQLAlchemy's event system and would like to be able to tell if a **query** originated from Core or ORM. My main interest is to audit everything that goes through SQLAlchemy and act based on it.

I know I can listen for ORM events, for example `after_flush`, and then inspect what happened to ORM entities; and I can listen for Core events, for example `after_execute`, and inspect executed queries and their parameters. However, if I'm not doing something wrong, queries from either Core and ORM do go through `after_execute`.

For example, in Python 3.9.6 and SQLAlchemy 1.4.11:

```python
from typing import Any

from sqlalchemy import Column, Integer, String, create_engine, event, insert
from sqlalchemy.engine import Connection, CursorResult
from sqlalchemy.orm import declarative_base, sessionmaker

Base = declarative_base()


class Person(Base):
    __tablename__ = "person"

    id = Column(Integer, primary_key=True)
    age = Column(Integer)
    name = Column(String(64), nullable=False)


def after_execute(
    conn: Connection,
    clauseelement: Any,
    multiparams: list[dict[str, Any]],
    params: dict[str, Any],
    execution_options: dict[str, Any],
    result: CursorResult,
) -> None:
    print(f"{clauseelement}\n{multiparams=}\n{params=}")


def test_mwe() -> None:
    engine = create_engine("sqlite+pysqlite:///db.sqlite3", echo=False, future=True)
    Base.metadata.create_all(engine)
    Session = sessionmaker(engine, future=True)
    event.listen(engine, "after_execute", after_execute)

    with engine.connect() as conn:
        conn.execute(insert(Person), dict(name="Someone", age=37))

    with Session.begin() as sess:
        sess.add(Person(name="Someone Else", age=43))
```

Outputs:

```
INSERT INTO person (id, age, name) VALUES (:id, :age, :name)
multiparams=[]
params={'name': 'Someone', 'age': 37}
INSERT INTO person (id, age, name) VALUES (:id, :age, :name)
multiparams=[]
params={'name': 'Someone Else', 'age': 43}
```

And I could not find a way to tell from any of the `after_execute` arguments if the `INSERT` was from Core or ORM. Is there any way to do this?

In a similar fashion, ORM-enabled `UPDATE`s and `DELETE`s are also intercepted. And there's a [warning](https://docs.sqlalchemy.org/en/14/orm/session_basics.html#selecting-a-synchronization-strategy) in the docs that goes:

>In order to intercept ORM-enabled UPDATE and DELETE operations with event handlers, use the [SessionEvents.do_orm_execute()](https://docs.sqlalchemy.org/en/14/orm/events.html#sqlalchemy.orm.SessionEvents.do_orm_execute) event.

So I don't know how to tell the difference from regular ORM updates and deletes and ORM-enabled ones, since both go through `do_orm_execute`.

---
Sidenote: I don't know why some queries are intercepted by `after_execute` but both `multiparams` and `params` come empty.
For example: change the Core insert above to `insert(Person).values(dict(name="Someone", age=37))` and re-run the code. This also happens for ORM-enabled updates and deletes. Is this behavior documented? What are the limitations that I would impose myself if I raised an exeption every time I received them empty?

**Comment by zzzeek** [comment]
so this still sort of depends on what you need, do_orm_execute() is triggered for all use of `session.execute()` as well as `session.query()` use, but wont give you unit of work queries.    All queries of course go through the connection-level events but when they are sent at that stage, they are all "core" queries, there are ways to "guess" where they came from, but nothing consistent if you are looking at both ORM SELECT queries as well as unit of work queries.  the most generalized, one-shot way to do that would be to use [frame introspection](https://docs.python.org/3/library/sys.html#sys._getframe) and look for modules with the prefix "sqlalchemy.orm" in the traceback.

>  I don't know why some queries are intercepted by after_execute but both multiparams and params come empty.
For example: change the Core insert above to insert(Person).values(dict(name="Someone", age=37)) and re-run the code. This also happens for ORM-enabled updates and deletes. Is this behavior documented?

the form of `.execute()` is `.execute(statement, parameters)`.  If you execute an `insert().values()`, that looks like `.execute(insert.values(dict(....)), None)`, so, there's no parameters there; if you are using events like `after_execute()`, you get those arguments as they were sent and [yes this is documented](https://docs.sqlalchemy.org/en/14/core/events.html?highlight=after_execute#sqlalchemy.events.ConnectionEvents.after_execute) as the arguments that were passed to the `.execute()` method; if you want what was **actually sent to the cursor**, then you use [after_cursor_execute](https://docs.sqlalchemy.org/en/14/core/events.html?highlight=after_execute#sqlalchemy.events.ConnectionEvents.after_cursor_execute).  For background on what the "cursor" is, see [pep 249](https://www.python.org/dev/peps/pep-0249/) which is linked all over SQLAlchemy's documentation as the underlying system we are calling into.

**Reply by ggabriel96** [reply]
Hey @zzzeek, thanks for your answer! So there is no way to tell the "origin" of the query via SQLAlchemy, right? Is that a feature of interest for the project? If so, I would consider contributing it :smile: 

As for the empty parameters topic... I just didn't understand exactly when and why is the case for empty (multi)parameters. Is it _only_ when doing an `insert(...).values(...)`? If so, I think I could live without it, since the docs also mention it's a less common idiom. The thing is that I wanted to save the uncompiled query and parameters in order to avoid dialect-specific formats.

By the way, if `insert(...).values(...)` don't "populate" these execute parameters, where do they go? Are there two "internal paths" that are used to pass parameters to the cursor?
